components/console_cmd_mqtt/console_mqtt.c

```diff
@@ -0,0 +1,445 @@
+{
+    if (mqtt_cfg.broker.address.uri) {
+        free((char *)mqtt_cfg.broker.address.uri);
+    }
+
+    if (mqtt_cfg.credentials.username) {
+        free((char *)mqtt_cfg.credentials.username);
+    }
+
+    if (mqtt_cfg.credentials.authentication.password) {
+        free((char *)mqtt_cfg.credentials.authentication.password);
+    }
+}
+
+
+static int do_mqtt_cmd(int argc, char **argv)
+{
+    int nerrors = arg_parse(argc, argv, (void **)&mqtt_args);
+    if (nerrors != 0) {
+        arg_print_errors(stderr, mqtt_args.end, argv[0]);
+        return 1;
+    }
+
+    if (mqtt_args.connect->count > 0) {
+        esp_mqtt_client_config_t mqtt_cfg;
+
+        if (client_handle != NULL) {
+            ESP_LOGI(TAG, "mqtt client already connected");
+            return 0;
+        }
+
+        if ((mqtt_args.username->count > 0) && (mqtt_args.password->count > 0)) {
+            if (mqtt_args.uri->count > 0) {
+                mqtt_cfg = get_mqtt_config_username_password(mqtt_args.uri->sval[0],
+                           mqtt_args.username->sval[0],
+                           mqtt_args.password->sval[0]);
+            } else {
+                mqtt_cfg = get_mqtt_config_username_password(CONFIG_MQTT_BROKER_URL,
+                           mqtt_args.username->sval[0],
+                           mqtt_args.password->sval[0]);
+            }
+        } else {
+            if (mqtt_args.uri->count > 0) {
+                mqtt_cfg = get_mqtt_config_basic(mqtt_args.uri->sval[0]);
+            } else {
+                mqtt_cfg = get_mqtt_config_basic(CONFIG_MQTT_BROKER_URL);
+            }
+        }
+
+        ESP_LOGI(TAG, "broker: %s", mqtt_cfg.broker.address.uri);
+
+        if (mqtt_args.cafile->count > 0) {
+            if (server_cert_pem_start && server_cert_pem_end) {
+                mqtt_cfg.broker.verification.certificate = (const char *)server_cert_pem_start;
+            }
+        }
+
+        if (mqtt_args.cert->count > 0) {
+            if (client_cert_pem_start && client_cert_pem_end) {
+                mqtt_cfg.credentials.authentication.certificate = (const char *)client_cert_pem_start;
+            }
+
+            if ((mqtt_args.key->count > 0) && client_key_pem_start && client_key_pem_end) {
+                mqtt_cfg.credentials.authentication.key = (const char *)client_key_pem_start;
+            } else {
+                mqtt_cfg.credentials.authentication.key = NULL;
+            }
+        }
+
+        client_handle = esp_mqtt_client_init(&mqtt_cfg);
+        if (client_handle == NULL) {
+            ESP_LOGE(TAG, "ERROR: Client init");
+            return 1;
+        }
+
+        /* The last argument may be used to pass data to the event handler, in this example mqtt_event_handler */
+        esp_mqtt_client_register_event(client_handle, ESP_EVENT_ANY_ID, mqtt_event_handler, NULL);
+        esp_mqtt_client_start(client_handle);
+
+        mqtt_config_free(mqtt_cfg);
```

Comment:
Note that we initialize the client in this function, so it's not necessary to create local copies of the common string fields, like URI, name, password...

Reply:
For name and password local copy was not required.
But for URI, without local heap or data allocation the application keeps crashing in `esp_mqtt_client_start()`.

Reply:
Resolved this issue. Thanks @david-cermak .